File: lightbulb-easyrecyclerview/src/main/java/com/github/rooneyandshadows/lightbulb/easyrecyclerview/layout_managers/VerticalFlowLayoutManager.java
package com.github.rooneyandshadows.lightbulb.easyrecyclerview.layout_managers;


import android.view.View;

import com.github.rooneyandshadows.lightbulb.easyrecyclerview.EasyRecyclerView;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterDataModel;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyRecyclerAdapter;
import com.google.android.flexbox.FlexDirection;
import com.google.android.flexbox.FlexboxLayoutManager;
import com.google.android.flexbox.JustifyContent;

import androidx.recyclerview.widget.RecyclerView;

public final class VerticalFlowLayoutManager<IType extends EasyAdapterDataModel, AType extends EasyRecyclerAdapter<IType>> extends FlexboxLayoutManager {
    private final EasyRecyclerView<IType, AType> easyRecyclerView;

    public VerticalFlowLayoutManager(EasyRecyclerView<IType, AType> easyRecyclerView) {
        super(easyRecyclerView.getContext(), FlexDirection.ROW);
        setJustifyContent(JustifyContent.FLEX_START);
        this.easyRecyclerView = easyRecyclerView;
    }

    @Override
    public int scrollVerticallyBy(int dy, RecyclerView.Recycler recycler, RecyclerView.State state) {
        int scrollRange = super.scrollVerticallyBy(dy, recycler, state);
        int overScroll = dy - scrollRange;
        if (Math.abs(dy) > 20)
            easyRecyclerView.getParent().requestDisallowInterceptTouchEvent(true);
        if (needToLoadMoreData(dy))
            handleLoadMore();
        return scrollRange;
    }

    private boolean needToLoadMoreData(int dy) {
        return (easyRecyclerView.hasMoreDataToLoad() &&
                !easyRecyclerView.isShowingLoadingHeader() &&
                !easyRecyclerView.isAnimating() &&
                !easyRecyclerView.isShowingRefreshLayout() &&
                !easyRecyclerView.isShowingLoadingFooter() &&
                dy > 0);
    }

    private void handleLoadMore() {
        if (!easyRecyclerView.supportsLazyLoading())
            return;
        View lastView = getChildAt(getChildCount() - 1);
        if (lastView == null)
            return;
        int size = easyRecyclerView.getItems().size();
        int last = ((RecyclerView.LayoutParams) lastView.getLayoutParams()).getAbsoluteAdapterPosition() - easyRecyclerView.getAdapter().getHeadersCount();
        if (last == size - 1)
            easyRecyclerView.loadMoreData();
    }
}

File: lightbulb-easyrecyclerview/src/main/java/com/github/rooneyandshadows/lightbulb/easyrecyclerview/layout_managers/VerticalFlowLayoutManager.kt
package com.github.rooneyandshadows.lightbulb.easyrecyclerview.layout_managers

import androidx.recyclerview.widget.RecyclerView
import com.github.rooneyandshadows.lightbulb.easyrecyclerview.EasyRecyclerView
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.data.EasyAdapterDataModel
import com.google.android.flexbox.FlexDirection
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.flexbox.JustifyContent
import kotlin.math.abs

class VerticalFlowLayoutManager<ItemType : EasyAdapterDataModel>(
    private val easyRecyclerView: EasyRecyclerView<ItemType>,
) : FlexboxLayoutManager(easyRecyclerView.context, FlexDirection.ROW) {

    init {
        justifyContent = JustifyContent.FLEX_START
    }

    @Override
    override fun scrollVerticallyBy(
        dy: Int,
        recycler: RecyclerView.Recycler,
        state: RecyclerView.State
    ): Int {
        //val overScroll = dy - scrollRange
        val scrollRange = super.scrollVerticallyBy(dy, recycler, state)
        if (abs(dy) > 20) {
            easyRecyclerView.parent.requestDisallowInterceptTouchEvent(true)
        }
        if (needToLoadData(dy)) {
            easyRecyclerView.loadMoreData()
        }
        return scrollRange
    }

    private fun needToLoadData(dy: Int): Boolean {
        val adapter = easyRecyclerView.adapter
        if (adapter == null || dy <= 0) return false
        val lastView = getChildAt(childCount - 1) ?: return false
        val lastViewAdapterPos =
            (lastView.layoutParams as RecyclerView.LayoutParams).absoluteAdapterPosition
        val headersCount = adapter.headersCount
        val visibleLastPosition = lastViewAdapterPos - headersCount
        val totalSize = adapter.collection.size()
        return visibleLastPosition == totalSize - 1
    }
}